### PR TITLE
Avoid panic() when the test fails.

### PR DESCRIPTION
**Pull request by jameinel**
## Description of change

If checkStartInstance fails with an Assert failure, we weren't
forcing the goroutine to exit, which meant it could get a panic() after
we had already started tearing down the test, which would then suppress
the real error that was occurring.
This forces us to kill the goroutine before we return from the test.

## QA steps

You can add this patch to force a failure:
```
--- a/worker/provisioner/provisioner_test.go
+++ b/worker/provisioner/provisioner_test.go
@@ -1555,6 +1555,7 @@ func (s *ProvisionerSuite) TestProvisionerRetriesTransientErrors(c *gc.C) {
                        }
                }
        }()
+       c.Fatalf("dead now")
        s.checkStartInstance(c, m3)
        close(thatsAllFolks)
        thatsAllFolks = nil
```

Without this patch, you get a failure like:
```
$ go test -check.v -check.f Retries
panic: Session already closed

goroutine 379 [running]:
gopkg.in/mgo%2ev2.(*Session).cluster(0xc4203dcd00, 0xc420749930)
        /home/jameinel/dev/go/src/gopkg.in/mgo.v2/session.go:1624 +0x56
gopkg.in/mgo%2ev2.copySession(0xc4203dcd00, 0x1, 0x0)
        /home/jameinel/dev/go/src/gopkg.in/mgo.v2/session.go:540 +0x40
gopkg.in/mgo%2ev2.(*Session).Copy(0xc4203dcd00, 0xc420749a50)
        /home/jameinel/dev/go/src/gopkg.in/mgo.v2/session.go:1590 +0x40
github.com/juju/juju/mongo.CollectionFromName(0xc420241d20, 0x18de255, 0xf, 0xf, 0xc4204888e0, 0x101)
        /home/jameinel/dev/go/src/github.com/juju/juju/mongo/collections.go:16 +0x32
github.com/juju/juju/state.(*database).GetCollection(0xc420233380, 0x18de255, 0xf, 0xc420749c28, 0xc48513, 0xc420749bc8)
        /home/jameinel/dev/go/src/github.com/juju/juju/state/database.go:282 +0x324
github.com/juju/juju/state.probablyUpdateStatusHistory(0x1a6dfe0, 0xc420233380, 0xc420380fc0, 0xc, 0x0, 0x0, 0x18e2f98, 0x12, 0x18cead9, 0x4, ...)
        /home/jameinel/dev/go/src/github.com/juju/juju/state/status.go:359 +0x123
github.com/juju/juju/state.setStatus(0x1a6dfe0, 0xc420233380, 0x18d3e90, 0x8, 0xc420380fc0, 0xc, 0x18e2f98, 0x12, 0x18cead9, 0x4, ...)
        /home/jameinel/dev/go/src/github.com/juju/juju/state/status.go:283 +0x210
github.com/juju/juju/state.(*Machine).SetInstanceStatus(0xc42059bd40, 0x18e2f98, 0x12, 0x18cead9, 0x4, 0xc4205ebbf0, 0xc4207fe800, 0x65b79dd79b8, 0x1d00000001)
        /home/jameinel/dev/go/src/github.com/juju/juju/state/machine.go:1055 +0x1e7
github.com/juju/juju/worker/provisioner_test.(*ProvisionerSuite).TestProvisionerRetriesTransientErrors.func1(0xc42004a420, 0xc42004a3c0, 0xc420278330, 0xc42059bd40, 0xc4201023c0)
        /home/jameinel/dev/go/src/github.com/juju/juju/worker/provisioner/provisioner_test.go:1542 +0x306
created by github.com/juju/juju/worker/provisioner_test.(*ProvisionerSuite).TestProvisionerRetriesTransientErrors
        /home/jameinel/dev/go/src/github.com/juju/juju/worker/provisioner/provisioner_test.go:1527 +0x62f
exit status 2
```

Which makes it impossible to see the original error. With this patch, you get the proper failure
```
[LOG] 0:01.473 TRACE juju.provisioner processMachinesWithTransientErrors([])
    c.Fatalf("dead now")
... Error: dead now
```

## Documentation changes

None.

## Bug reference

We have lot's of bugs related to `panic()` etc in this function, some of which we might have fixed, but it is hard to know, because the panic hides the real error.

 * https://bugs.launchpad.net/juju/+bug/1394223
 * https://bugs.launchpad.net/juju/+bug/1749709
 * https://bugs.launchpad.net/juju/+bug/1753407
 * https://bugs.launchpad.net/juju/+bug/1754021

### CODE REVIEW

File: worker/provisioner/provisioner_test.go [hunk 1]
```diff
@@ -1524,6 +1524,17 @@ func (s *ProvisionerSuite) TestProvisionerRetriesTransientErrors(c *gc.C) {
 	// instance has started.
 	thatsAllFolks := make(chan struct{})
```
Comment (manadart) [inline worker/provisioner/provisioner_test.go]: Worth renaming to _m3StartConfirmed_ or some such? I found I definitely had to read the rest of the test to assign meaning before going back and continuing to parse.

File: worker/provisioner/provisioner_test.go [hunk 2]
```diff
@@ -1524,6 +1524,17 @@ func (s *ProvisionerSuite) TestProvisionerRetriesTransientErrors(c *gc.C) {
...
 	hackyGoroutineDone := make(chan struct{})
```
Comment (manadart) [inline worker/provisioner/provisioner_test.go]: Same here.

File: worker/provisioner/provisioner_test.go (hunk 1 above)
Comment (jameinel) [reply worker/provisioner/provisioner_test.go]: I renamed them to runSetStatusGoroutine and setStatusGoroutineDone

Does that work better for you?

**Comment by jameinel** [comment]
$$merge$$

**Pull request merged**